### PR TITLE
[RFC] Automated Validation

### PR DESCRIPTION
**Pull request by nholland94**
[Link to RFC contents for convenience](https://github.com/CodaProtocol/coda/blob/rfc/automated-validation/rfcs/0032-automated-validation.md)

### CODE REVIEW

File: rfcs/0032-automated-validation.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+#### Languages Not Chosen
+
+With the requirement that this system be highly concurrent (so that we don't need to spend a lot of engineering effort trying to optimize this tool), that immediately hurts the case for the main language we use currently, OCaml. OCaml comes with great benefits, such as it already being our standard language on the team and it having a good static type system, but the single threaded concurrency story in OCaml is poor, and the solutions and tooling around them (rpc\_parallel) are bulky to use and quite difficult to iterate on. As such, OCaml seems like the wrong choice for this system as we would spend a lot of engineering effort on just the architecture for this system (and would likely need to revisit that as we intend to scale this system up to more and more validations and larger network sizes).
```
Comment (psteckler) [inline rfcs/0032-automated-validation.md:66]: Well, there's good-ish old system threads in OCaml, which you get by linking a library. It doesn't map to multiple cores, though. It's much lighter-weight than `rpc_parallel`.

Comment (nholland94) [reply rfcs/0032-automated-validation.md:66]: I assume you are referring to https://caml.inria.fr/pub/docs/manual-ocaml/libref/Thread.html ? It's certainly lighter-weight compared to `rpc_parallel`, but that comes at the cost of usability since we would need to write our own tooling for thread management and architect our own concurrent system, which seems like a lot of work.

File: rfcs/0032-automated-validation.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+#### Statistics
+[detailed-design-architecture-statistics]: #detailed-design-architecture-statistics
+
+Statistics consume network state information from one or more providers and incrementally compute a rolling state (representing the computed statistic). Statistics are (at least in the initial design) individually scoped to resources (meaning that each statistic has a singular instance computed for each resource that is being monitored by the system). Statistics broadcast their state updates so that vlaidations can be performed on one or more statistic as values change. Statistics can also (if desired) be forwarded to as prometheus metrics (though this is not a hard requirement of the design and likely will be omitted from the first pass). Doing this would allow us to compute more meaningful network wide statistics that we cannot easily compute today without writing a program to scrape the logs after the network runs.
```
Comment (bkase) [inline rfcs/0032-automated-validation.md:47]: Can you use more decisive language here?

> Statistics can also (if desired) be forwarded to as prometheus metrics (though this is not a hard requirement of the design and likely will be omitted from the first pass). Doing this would allow us to compute more meaningful network wide statistics that we cannot easily compute today without writing a program to scrape the logs after the network runs.

This could be reworded like:

> While useful, explicitly out of scope for the initial pass of the project is forwarding statistics as Prometheus metrics. Doing this...

File: rfcs/0032-automated-validation.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+## Prior art
+[prior-art]: #prior-art
+
+- Conner has written a few Python scripts in the `coda-automation` repo which we currently use to generate graphics that we manually validate
+  - much of this work will probably be ported into this tool at some point, so this work is still important and provides a basis for some of this services functionality
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- is the protocol team comfortable with maintaining a new service that is not written in OCaml?
```
Comment (deepthiskumar) [inline rfcs/0032-automated-validation.md:218]: +1

File: rfcs/0032-automated-validation.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+- credentials to talk to various APIs (gcloud, kubernetes, discord, etc...)
+- information about which network to validate
+- the set of resources deployed on the network (see below for more detail on resources)
+- a list of "validation queries"; each "validation query" includes:
+  - the name of the validation to perform
+  - a query on which resources it should be performed on
+
+From the list of validation queries, the automated validation can solve for other information it needs to actually run, such as what log filters it needs to pull from stackdriver, what information to compute from events and state changes on the network, and what resources are scoped to each of these concepts. The architecture of the automated validation service is thus broken up and modeled as a series of concurrent "processes", which we will classify into various roles, which are responsible for one piece of the entire service.
```
[the remark this answers is not in the document]
Comment (deepthiskumar) [reply rfcs/0032-automated-validation.md:30]: A diagram would be nice. The way I understood is: 

     Resources <--events-- Provider <--raw-results-- Consumer (The statistician) <--accumulated-results-- Validator --alerts--> External entity (discord, integration test)

(The arrows signify where we look for the data)

File: rfcs/0032-automated-validation.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+The primary motivations behind building an automated validation service are to reduce validation costs to engineering time, tighten the iteration loop around stabilizing our network, give us more confidence in future public testnet releases, and to provide a route to regular CI network validations.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+Observing the [list of validations we require for stable networks](https://www.notion.so/codaprotocol/Testnet-Validation-1705d5bf03e04d03b21a1f5724a17597), the majority of validations can be answered through a combination of log processing and graphql queries. As such, we will initially scope the design of the validation service to focus on those two aspects, while trying to create a design that isn't over-fitted to these two aspects so that we can refit parts of the architecture for validations which are outside this scope after the initial implementation is complete.
```
[the remark this answers is not in the document]
Comment (nholland94) [reply rfcs/0032-automated-validation.md:16]: Yes. I will make the Notion page public now, but I think it makes the most sense to open a new PR that commits the list of validations to the repository somewhere.

File: rfcs/0032-automated-validation.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (nholland94) [reply rfcs/0032-automated-validation.md:30]: I will make a quick diagram to show the data flow between providers, statistics, and validations.

File: rfcs/0032-automated-validation.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+#### Resources
+[detailed-design-architecture-resources]: #detailed-design-architecture-resources
+
+A deployment consists of a series of pods and containers. We define a universal abstraction for reasoning about these artifacts of a deployment, which generically call "resources". Resources are classified into different categories in an inheritance model (similar to OOP in structure, but this is not talking about OOP classes; no methods, mutability, or other OOP concepts). For example, the main resource we are interested in is a `CodaNode`. There are multiple subclasses of `CodaNode`, such as `BlockProducer`, `SnarkCoordinator`, or `Seed`. Each of these classes may have different metadata associate with instances of them. For instance, all `CodaNode` instances have a `pod_name`, and `BlockProducer` instances have additional metadata for what block producer role they are and their integer id (eg in "fish-block-producer-1", role="fish" and id=1).
```
[the remark this answers is not in the document]
Comment (nholland94) [reply rfcs/0032-automated-validation.md:35]: Capabilities are defined by how the resources are used, not by the resources. The resources are really just a subtyping relationship around some metadata that identifies an individual resource. I will rephrase this as such.

File: rfcs/0032-automated-validation.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+[prior-art]: #prior-art
+
+- Conner has written a few Python scripts in the `coda-automation` repo which we currently use to generate graphics that we manually validate
+  - much of this work will probably be ported into this tool at some point, so this work is still important and provides a basis for some of this services functionality
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- is the protocol team comfortable with maintaining a new service that is not written in OCaml?
+- should this service be used as part of the integration testing framework (there is a large overlap in what and how this validation service will monitor the health of the network and what the new integration test framework project intends to achieve)
```
[the remark this answers is not in the document]
Comment (nholland94) [reply rfcs/0032-automated-validation.md:219]: We may also have to expand the scope of this to support dynamic validations being enabled/disabled from the test executive. Probably wouldn't be too hard, but it does add some extra overhead to utilizing this for the integration tests. I still believe it's worth it though.

File: rfcs/0032-automated-validation.md (hunk 2 above)
Comment (nholland94) [reply rfcs/0032-automated-validation.md:47]: After talking about the utility of this with Conner and Aneesha at various points over the last week, I actually think we should just do this. It's not too much more work and has a high benefit. I'm changing the language here to commit to it.

File: rfcs/0032-automated-validation.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+#### Statistics
+[detailed-design-architecture-statistics]: #detailed-design-architecture-statistics
+
+Statistics consume network state information from one or more providers and incrementally compute a rolling state (representing the computed statistic). Statistics are (at least in the initial design) individually scoped to resources (meaning that each statistic has a singular instance computed for each resource that is being monitored by the system). Statistics broadcast their state updates so that vlaidations can be performed on one or more statistic as values change. Statistics can also be forwarded to as prometheus metrics. Doing this would allow us to compute more meaningful network wide statistics that we cannot easily compute today without writing a program to scrape the logs after the network runs.
```
Comment (emberian) [inline rfcs/0032-automated-validation.md:55]: https://joshday.github.io/OnlineStats.jl/v0.10/pages/types.html we can algorithms from this as we need them I guess. also, typo, vlaidations

**Comment by emberian** [comment]
I'm somewhat skeptical of complicating our tech stack _even more_, but I agree that elixir is a great fit for this service and I think it would be more pain to build out the ocaml libraries we'd need than to just use elixir. I'm in favor of this plan!

File: rfcs/0032-automated-validation.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+Below is a diagram which displays how the system is initialized given the validation queries and resource database. From the information computed during this initialization, the automated validation service can create the various processes and external resources (such as log sinks on google cloud) necessary to perform the requested validations on the requested resources.
+
+![](./res/automated_validation_init_architecture.conv.tex.png)
+
+The following diagram shows the runtime architecture of the processes in the system once initialization has been completed. Note how this diagram shows that statistics may consume multiple providers, and validations may consume multiple statistics.
+
+![](./res/automated_validation_runtime_architecture.conv.tex.png)
```
Comment (bkase) [inline rfcs/0032-automated-validation.md:38]: This is _so_ useful, thanks!
